Include aspect-ratio when calculating inline min-content size and add min-content block computation

sammygill opened the pull request:
#### 588e20666a36f28117f49082462513f03e25960e
<pre>
Include aspect-ratio when calculating inline min-content size and add min-content block computation
<a href="https://bugs.webkit.org/show_bug.cgi?id=243473">https://bugs.webkit.org/show_bug.cgi?id=243473</a>

Reviewed by Rob Buis.

When we compute the content size suggestion of an element with a
preferred aspect ratio, we need to make sure that the minimum-content
size of the element is computed through the aspect ratio.
When the main axis is in the inline direction, we have to consider
whether or not an aspect-ratio is specified. If there is an aspect-ratio,
we must calculate the minimum-content size using that. Otherwise, we
will just compute the main axis extent for the child.

However, the definition of of the minimum-content size in the block
direction is slightly different. According to the CSS Sizing spec,
the min-content block size is: he box&apos;s &quot;ideal&quot; size in the block axis.
Usually the block size of the content after layout.

* LayoutTests/TestExpectations:
* Source/WebCore/rendering/RenderBox.cpp:
(WebCore::inlineSizeFromAspectRatio):

* Source/WebCore/rendering/RenderFlexibleBox.cpp:
(WebCore::RenderFlexibleBox::computeFlexItemMinMaxSizes):
(WebCore::RenderFlexibleBox::adjustChildSizeForAspectRatioCrossAxisMinAndMax):

Canonical link: <a href="https://commits.webkit.org/253740@main">https://commits.webkit.org/253740@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/5704bc1b58ccb3970f3c9f0248b032d3f84ca110

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/86848 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/69/builds/30931 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/17731 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/95679 "Built successfully") | [✅ 🛠 🧪 win](https://ews-build.webkit.org/#/builders/10/builds/149434 "Built successfully and passed tests") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/90828 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/64/builds/29292 "Built successfully") | [✅ 🛠 mac-debug](https://ews-build.webkit.org/#/builders/71/builds/25649 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/79006 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/90912 "Built successfully") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/92464 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/68/builds/23654 "Passed tests") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/73723 "Built successfully") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/23685 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/9/builds/78649 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/78944 "Passed tests") | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🧪 api-gtk](https://ews-build.webkit.org/#/builders/34/builds/66690 "Build is in progress") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/66/builds/27051 "Built successfully") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/73/builds/12797 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/67/builds/26974 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/13811 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/2652 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/63/builds/28658 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/60/builds/36677 "Passed tests") | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/65/builds/28602 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/33090 "Passed tests") | | 
<!--EWS-Status-Bubble-End-->

rwlbuis reviewed (commented):
There are quite a number of typos to fix (minimuim etc.).

Overall progress looks great. However flex-aspect-ratio-026.html is not completely fixed. I worry a bit that improving/fixing that may need touching this code section again, do you have any idea about that?

rwlbuis commented on Source/WebCore/rendering/RenderBox.cpp:
```diff
@@ -2721,7 +2721,7 @@ void RenderBox::updateLogicalWidth()
 {
     if (boxSizing == BoxSizing::BorderBox)
-        return std::max(borderPaddingInlineSum, LayoutUnit(blockSize * aspectRatio));
+        return LayoutUnit(blockSize * aspectRatio);
```
Can you explain this change?

rwlbuis commented on Source/WebCore/rendering/RenderFlexibleBox.cpp:
```diff
@@ -1390,11 +1391,22 @@ std::pair<LayoutUnit, LayoutUnit> RenderFlexibleBox::computeFlexItemMinMaxSizes(
-            contentSize = computeMainSizeFromAspectRatioUsing(child, childCrossSizeLength);
+        
+        LayoutUnit sizeThroughAspectRatio;
+        LayoutUnit minContentSize = computeMainAxisExtentForChild(child, MinSize, Length(LengthType::MinContent)).value_or(0);
```
This should probably be a lambda, since one code path below does not need it.

rwlbuis reviewed (approved):
LGTM.

sammygill committed:
Include aspect-ratio when calculating inline min-content size and add min-content block computation
https://bugs.webkit.org/show_bug.cgi?id=243473

Reviewed by Rob Buis.

When we compute the content size suggestion of an element with a
preferred aspect ratio, we need to make sure that the minimum-content
size of the element is computed through the aspect ratio.
When the main axis is in the inline direction, we have to consider
whether or not an aspect-ratio is specified. If there is an aspect-ratio,
we must calculate the minimum-content size using that. Otherwise, we
will just compute the main axis extent for the child.

However, the definition of of the minimum-content size in the block
direction is slightly different. According to the CSS Sizing spec,
the min-content block size is: he box's "ideal" size in the block axis.
Usually the block size of the content after layout.

* LayoutTests/TestExpectations:
* Source/WebCore/rendering/RenderBox.cpp:
(WebCore::inlineSizeFromAspectRatio):

* Source/WebCore/rendering/RenderFlexibleBox.cpp:
(WebCore::RenderFlexibleBox::computeFlexItemMinMaxSizes):
(WebCore::RenderFlexibleBox::adjustChildSizeForAspectRatioCrossAxisMinAndMax):

Canonical link: https://commits.webkit.org/253740@main
